### PR TITLE
fix: improve handling optimization profiles

### PR DESCRIPTION
**Pull request by junstar92**
- Remove the constraint on multiples of 2 or 8 for dynamic dimensions
- Support the dynamic dimensions with the same min and max values

Note: Some constants of dynamic dimensions are not instances of `DynamicDimension`, but simply `Number`. These values might be updated in the future as new features are added.

### CODE REVIEW

File: src/ditto/arguments/dynamic_dim.py [hunk 1]
```diff
@@ -521,3 +515,31 @@ def check_at_least_one_of_lhs_or_rhs_is_dynamic_dim(self) -> Self:
...
+    expr = Symbol(f"s{len(shape_env.var_to_val)}", positive=True, integer=True)
+    shape_env.var_to_val[expr] = Integer(val)
+    shape_env.var_to_range[expr] = constraint_dim.vr
+    shape_env.counter["create_symbol"] += 1
+    r_sources = shape_env.var_to_sources[expr] = []
+    r_sources.append(source)
+    if not source.is_ephemeral() and r_sources[0].is_ephemeral():
+        r_sources[0], r_sources[-1] = r_sources[-1], r_sources[0]
+    shape_env.symbol_guard_counter[expr] = 0
+    return SymInt(SymNode(expr, shape_env, int, val))
```
Comment (jiwoong-choi) [inline src/ditto/arguments/dynamic_dim.py:545]: Creating `torch.SymInt` like this can cause a problem. I used to use a similar approach before to create a `torch.SymInt` object, but replaced it with a bit higher-level APIs provided by `ShapeEnv` like in [`DynamicDimensionType._create_sym_int`](https://github.com/SqueezeBits/Torch-TRTLLM/blob/dceacfbeb074b49e6cccc31cbf6d204feed10b1e/src/ditto/arguments/dynamic_dim.py#L113-L128) method. (Cannot remember the reason, but it caused a problem at some point ...)
Would it be possible to eliminate lines like
```
    shape_env.var_to_val[expr] = Integer(val)
    shape_env.var_to_range[expr] = constraint_dim.vr
    shape_env.counter["create_symbol"] += 1
    r_sources = shape_env.var_to_sources[expr] = []
```
that manually set the properties of `ShapeEnv` and replace them with the higher-level APIs provided by `ShapeEnv`?
Or, at least, create a `torch.SymInt` object with the higher-level API and modify some of its attributes afterward, as there are tonnes of other attributes of `ShapeEnv` not properly handled in this way.

File: src/ditto/arguments/torch_export_arguments.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def from_hints(
                     continue
                 assert (constraint := constraints[name]) is not None
                 if not isinstance(export_dim := s.export_dim, int):
-                    constraint[dim] = export_dim
+                    constraint[dim] = 2 * export_dim
```
Comment (jiwoong-choi) [inline src/ditto/arguments/torch_export_arguments.py:97]: Why do we need `2 *` here?

File: src/ditto/arguments/trtllm_argument_hint.py [hunk 3]
```diff
@@ -124,6 +125,23 @@ def as_dict(self) -> dict[str, TensorTypeHint | None]:
         """
         return TypeAdapter(dict[str, TensorTypeHint | None]).validate_python(self.model_dump())
 
+    def create_dynamic_dim(self, name: str, ranges: list[int]) -> DynamicDimension:
```
Comment (jiwoong-choi) [inline src/ditto/arguments/trtllm_argument_hint.py:128]: Why is `range` of type `list[int]` while the method `create_dynamic_dim` only receives lists filled with only one integer?

File: src/ditto/configs/trtllm/optimization_profile.py [hunk 4]
```diff
@@ -118,17 +124,17 @@ def create_from(
             opt_num_tokens = min(max_num_tokens, max_batch_size * max_beam_width)
             logger.debug(f"opt_num_tokens is not set, specifying to {opt_num_tokens}")
         max_kv_cache_block_size = math.ceil(max_seq_len / plugin_config.tokens_per_block)
-        opt_kv_cache_block_size = math.ceil((max_seq_len // 2) / plugin_config.tokens_per_block)
+        opt_kv_cache_block_size = math.ceil(divide_by_2(max_seq_len) / plugin_config.tokens_per_block)
         return cls(
             max_batch_size=max_batch_size,
-            opt_batch_size=(max_batch_size + 1) // 2,
+            opt_batch_size=divide_by_2(max_batch_size, offset=1),
```
Comment (jiwoong-choi) [inline src/ditto/configs/trtllm/optimization_profile.py:130]:
```suggestion
            opt_batch_size=max(1, (max_batch_size + 1) // 2),
```

File: src/ditto/arguments/dynamic_dim.py (hunk 1 above)
Comment (junstar92) [reply src/ditto/arguments/dynamic_dim.py:545]: I attempted to use the APIs provided by PyTorch, but they all seem to return an `Int` instead of `SymNode` when the min and max values are the same. As a result, directly creating a `SymNode` appears to be the only viable approach.
I agree that manually modifying the properties of `ShapeEnv` comes with risks. While it may be possible to generate only the `SymNode` without modifying the attributes, I'm not entirely sure how `ShapeEnv` interacts with `torch.export` and torch's `FakeMode`. For this reason, I chose to replicate the behavior of the APIs of `ShapeEnv` as closely as possible.

If we find that modifying the properties isn't necessary and doesn't cause errors, we could consider just creating `SymNode` directly without those modifications. However, further testing is required. I'll run this through multiple models through our test script and attach the results for review.

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (junstar92) [reply src/ditto/arguments/torch_export_arguments.py:97]: It's to handle the constraints(e.g., multiple of 2) imposed by some layers like split layer.

The `export_dim` used here is actually a range intended for TensorRT's optimization profiles, so when used in `torch.export`, I believe this trick doesn't really matter. This is solely for extracting the graph, and the dynamic dimension's range is only needed during the TensorRT build.

However, it might be worth examining how a large range could potentially impact `torch.export`. Do you know if having a large max value for the dynamic dimension constraint affects the export time? 
(From the tests I ran on a few models, it doesn't seem to have any noticeable impact on export performance, but further confirmation would be helpful. I am also going to test it and attach the results.

File: src/ditto/arguments/trtllm_argument_hint.py (hunk 3 above)
Comment (junstar92) [reply src/ditto/arguments/trtllm_argument_hint.py:128]: Currently, there are no cases where the min, opt, and max values differ, but it was designed this way to handle potential cases like that in the future.

File: src/ditto/arguments/dynamic_dim.py (hunk 1 above)
Comment (junstar92) [reply src/ditto/arguments/dynamic_dim.py:545]: Some issues are occurring in certain cases during the `torch.export` or the optimization process. However, these issues seem to be separate from `ShapeEnv`. I will reopen the PR after finding a solution, and if possible, revert the modification to the properties of `ShapeEnv`.

**Comment by junstar92** [comment]
I will reopen after finding a better solution.

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (junstar92) [reply src/ditto/arguments/torch_export_arguments.py:97]: Test results have shown that `2 * ` is required to build a model with split layers like Phi-3.5.
But, is multiple of 8 better for potential cases?

File: src/ditto/arguments/trtllm_argument_hint.py (hunk 3 above)
Comment (junstar92) [reply src/ditto/arguments/trtllm_argument_hint.py:128]: I will make it more clearly.

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (jiwoong-choi) [reply src/ditto/arguments/torch_export_arguments.py:97]: I agree, the min/max values of a `torch.export.Dim` won't make any difference in the exported graph, and we don't need to care about the performance of the `ExportedProgram` or `GraphModule` as they are just IR required for graph-level optimizations for use.

File: src/ditto/arguments/trtllm_argument_hint.py (hunk 3 above)
Comment (jiwoong-choi) [reply src/ditto/arguments/trtllm_argument_hint.py]: Then it could be clearer to have `ranges: int | tuple[int, int, int]` as a function parameter and declare the local variable `dim_range` as `dim_range = (ranges, ranges, ranges) if isinstance(ranges, int) else ranges` .
(`list[int]` can have arbitrary number of elements, while you can also enforce the number of elements to 3 by giving the type hint as `tuple[int, int, int]`)

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (jiwoong-choi) [reply src/ditto/arguments/torch_export_arguments.py:97]: > Test results have shown that `2 * ` is required to build a model with split layers like Phi-3.5. Also, it might not have an impact on export performance. However, is multiple of 8 better for potential cases?

Well, we will never know how many splits a split layer will produce. If we want to be extra safe, then a scale factor like 24 or 60 would be a good choice, as they have many divisors. And this scale factor must be specified in `constants.py` for flexibility when there's a split layer producing an extraordinary number of splits, for example, 7 or 13.

File: src/ditto/arguments/trtllm_argument_hint.py [hunk 5]
```diff
@@ -53,6 +53,8 @@ class TRTLLMArgumentHint(StrictlyTyped):
     tp_size: int = Field(default=1, exclude=True, gt=0)
     gather_context_logits: bool = Field(default=False, exclude=True)
     lora_input_hints: dict[str, TensorTypeHint] = Field(default_factory=dict, exclude=True)
+    _one: DynamicDimension = PrivateAttr(default=DynamicDimension(name="one", min=1, opt=1, max=1))
+    _two: DynamicDimension = PrivateAttr(default=DynamicDimension(name="two", min=2, opt=2, max=2))
```
Comment (jiwoong-choi) [inline src/ditto/arguments/trtllm_argument_hint.py:57]: It looks like `_two` is never used. Let's remove it for now and add it back if it is needed later.

File: src/ditto/arguments/trtllm_argument_hint.py [hunk 6]
```diff
@@ -267,3 +298,18 @@ def serialize_model(
         data = original_serializer(self)
         data.update(self.lora_input_hints)
         return data
+
+
+def create_dynamic_dim(name: str, *, min_val: int, opt_val: int, max_val: int) -> DynamicDimension:
```
Comment (jiwoong-choi) [inline src/ditto/arguments/trtllm_argument_hint.py:303]: Can we just use the constructor of `DynamicDimension` instead of this utility function? It's an over-refactoring in my point of view.

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (junstar92) [reply src/ditto/arguments/torch_export_arguments.py:97]: Then, I think that there isn't a perfect solution and such cases are rare, so a scale factor of 2 might be sufficient for now.

File: src/ditto/arguments/trtllm_argument_hint.py (hunk 3 above)
Comment (junstar92) [reply src/ditto/arguments/trtllm_argument_hint.py]: I resolved it by removing this function and using the type directly.

File: src/ditto/arguments/torch_export_arguments.py (hunk 2 above)
Comment (kkimmk) [reply src/ditto/arguments/torch_export_arguments.py:97]: It seems the reason for `2 *` is similar to `8 *` in [the previous version](https://github.com/SqueezeBits/Torch-TRTLLM/blob/main/src/ditto/arguments/trtllm_argument_hint.py#L92). (correct me if I'm wrong). Just out of curiosity, if it's for the similar purpose, is there any reason for moving the code to this file rather than trtllm_argument_hint.py as in the previous code? If there's no difference, I think the previous location seems better because it's more related to the TensorRT than `torch.export`. Also, it'll be good to add a NOTE in the code to explain the purpose.

Comment (junstar92) [reply src/ditto/arguments/torch_export_arguments.py:97]: @kkimmk The reason for `2 *` is the same as for `8 * s`. But, in the previous version, this constraint(`8 * s`) affects optimization profiles for building TensorRT Engine despite it is completely unrelated to the optimization profiles. For example, we cannot set `opt_num_tokens` to a value less than 8, but it should technically be possible.

Also, as you mentioned, since `TRTLLMArgumentHint` is more related to TensorRT, it seems more natural to include the restrictions on `torch.export` in `TorchExportArguments` (`from_hint`), not `TRTLLMArgumentHint`.

**Review by kkimmk — APPROVED** [review]
